File: RobotCode/snobot2017/src/com/snobot2017/gearboss/SnobotGearBoss.java
package com.snobot2017.gearboss;

import com.snobot.lib.logging.ILogger;
import com.snobot2017.SmartDashBoardNames;
import com.snobot2017.joystick.IOperatorJoystick;

import edu.wpi.first.wpilibj.DoubleSolenoid;
import edu.wpi.first.wpilibj.DoubleSolenoid.Value;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

/**
 * Controls the gear boss
 * @author Owner/Nora
 *
 */
public class SnobotGearBoss implements IGearBoss
{
    private static final Value sGEAR_DOWN_VALUE = Value.kForward;
    private static final Value sGEAR_UP_VALUE = Value.kReverse;

    private DoubleSolenoid mGearSolenoid;
    private IOperatorJoystick mOperatorJoystick;
    private ILogger mLogger;
    private boolean mGearIsUp;

    public SnobotGearBoss(DoubleSolenoid aGearSolenoid, IOperatorJoystick aOperatorJoystick, ILogger aLogger)
    {
        mGearSolenoid = aGearSolenoid;
        mOperatorJoystick = aOperatorJoystick;
        mLogger = aLogger;

        moveGearHigh();
    }

    @Override
    public void init()
    {
        mLogger.addHeader("SolenoidPosition");
    }

    @Override
    public void update()
    {
        mGearIsUp = mGearSolenoid.get() == sGEAR_UP_VALUE;
    }

    @Override
    public void control()
    {
        switch (mOperatorJoystick.moveGearBossToPosition())
        {
        case UP:
        {
            moveGearHigh();
            break;
        }
        case DOWN:
        {
            moveGearLow();
            break;
        }
        case NONE:
        {
            break;
        }
        }
    }

    @Override
    public void rereadPreferences()
    {
        // Nothing
    }

    @Override
    public void updateSmartDashboard()
    {
        SmartDashboard.putBoolean(SmartDashBoardNames.sGEAR_BOSS_SOLENOID, mGearIsUp);
    }

    @Override
    public void updateLog()
    {
        mLogger.updateLogger(mGearIsUp);
    }

    @Override
    public void stop()
    {
        moveGearHigh();
    }

    @Override
    public void moveGearHigh()
    {
        mGearSolenoid.set(sGEAR_UP_VALUE);
    }

    @Override
    public void moveGearLow()
    {
        mGearSolenoid.set(sGEAR_DOWN_VALUE);
    }

    @Override
    public boolean isGearUp()
    {
        return mGearIsUp;
    }


}

File: Simulator/TestCpp/src/Gearboss/SnobotGearBoss.cpp
/*
 * SnobotGearBoss.cpp
 *
 *  Created on: May 19, 2017
 *      Author: preiniger
 */

#include "Gearboss/SnobotGearBoss.h"
#include "SmartDashboardNames.h"

// WpiLib
#include "SmartDashboard/SmartDashboard.h"


const DoubleSolenoid::Value SnobotGearBoss::sGEAR_DOWN_VALUE = DoubleSolenoid::kForward;
const DoubleSolenoid::Value SnobotGearBoss::sGEAR_UP_VALUE = DoubleSolenoid::kReverse;

SnobotGearBoss::SnobotGearBoss(std::shared_ptr<DoubleSolenoid> aGearSolenoid,
        std::shared_ptr<IOperatorJoystick> aOperatorJoystick,
        std::shared_ptr<ILogger> aLogger) :
        mGearIsUp(false)
{
    mGearSolenoid = aGearSolenoid;
    mOperatorJoystick = aOperatorJoystick;
    mLogger = aLogger;

    moveGearHigh();
}

SnobotGearBoss::~SnobotGearBoss()
{

}

void SnobotGearBoss::initializeLogHeaders()
{
    mLogger->addHeader("SolenoidPosition");
}

void SnobotGearBoss::update()
{
    mGearIsUp = mGearSolenoid->Get() == sGEAR_UP_VALUE;
    mOperatorJoystick->setShouldRumble(!mGearIsUp);
}

void SnobotGearBoss::control()
{
    switch (mOperatorJoystick->moveGearBossToPosition())
    {
    case IOperatorJoystick::GearBossPositions::UP:
    {
        moveGearHigh();
        break;
    }
    case IOperatorJoystick::GearBossPositions::DOWN:
    {
        moveGearLow();
        break;
    }
    case IOperatorJoystick::GearBossPositions::NONE:
    {
        break;
    }
    }
}

void SnobotGearBoss::updateSmartDashboard()
{
    SmartDashboard::PutBoolean(SmartDashboardNames::sGEAR_BOSS_SOLENOID,
            mGearIsUp);
}

void SnobotGearBoss::updateLog()
{
    mLogger->updateLogger(mGearIsUp);
}

void SnobotGearBoss::stop()
{
    moveGearHigh();
}

void SnobotGearBoss::moveGearHigh()
{
    mGearSolenoid->Set(sGEAR_UP_VALUE);
}

void SnobotGearBoss::moveGearLow()
{
    mGearSolenoid->Set(sGEAR_DOWN_VALUE);
}

bool SnobotGearBoss::isGearUp()
{
    return mGearIsUp;
}
